src/execution/value_stack.rs

```diff
@@ -12,81 +13,147 @@ use crate::unreachable_validated;
+
+    /// Get a shared reference to the current [CallFrame]
+    pub fn current_stackframe(&self) -> &CallFrame {
+        self.frames.last().unwrap_validated()
+    }
+
+    /// Get a mutable reference to the current [CallFrame]
+    pub fn current_stackframe_mut(&mut self) -> &mut CallFrame {
+        self.frames.last_mut().unwrap_validated()
     }
 
-    #[allow(warnings)]
-    /// Assumes this entry to be an [ActivationFrame]
-    pub fn into_activation(self) -> ActivationFrame {
-        match self {
-            StackEntry::Activation(a) => a,
-            _ => unreachable_validated!(),
+    /// Pop a [CallFrame] from the call stack
+    // TODO figure out if this has to change the value stack
+
+    pub fn pop_stackframe(&mut self) -> usize {
+        // TODO maybe manipulate the value stack
+        let stack_vals = self.frames.last().unwrap_validated().stack_values;
+        let return_addr = self.frames.pop().unwrap_validated().return_addr;
+
+        // Move ownership of previous stackframe's values to the previous stackframe
+        // In theroy, this should be only the return values of the previous stack frame.
+        if let Some(callframe) = self.frames.last_mut() {
+            callframe.stack_values += stack_vals;
         }
+
+        return_addr
+    }
+
+    /// Pop a stackframe from the call stack
+    // TODO figure out if this has to change the value stack
+    pub fn push_stackframe(&mut self, func_idx: FuncIdx, locals: Locals, return_addr: usize) {
+        // TODO maybe manipulate the value stack
+        self.frames.push(CallFrame {
+            func_idx,
+            locals,
+            return_addr,
+            stack_values: 0,
+        })
+    }
+
+    /// Returns how many stackframes are on the stack, in total.
+    pub fn callframe_count(&self) -> usize {
+        self.frames.len()
+    }
+
+    /// Pop `n` elements from the value stack's tail as an iterator, with the first element being
+    /// closest to the bottom of the value stack
+    pub fn pop_tail_iter(&mut self, n: usize) -> Drain<Value> {
+        let start = self.values.len() - n;
+        self.current_stackframe_mut().stack_values -= n;
+
+        self.values.drain(start..)
+    }
+
+    /// Clear all of the values pushed to the value stack by the current stack frame
+    pub fn clear_callframe_values(&mut self) {
+        self.pop_tail_iter(self.current_stackframe().stack_values);
+        self.current_stackframe_mut().stack_values = 0;
     }
 }
+
+/// The [WASM spec](https://webassembly.github.io/spec/core/exec/runtime.html#stack) calls this `Activations`, however it refers to the call frames of functions.
+pub(crate) struct CallFrame {
+    /// Index to the function of this CallFrame
+    pub func_idx: FuncIdx,
+
+    /// Local varaiables such as parameters for this [CallFrame]'s function
+    /// TODO(george-cosma): Do we want "locals" to be a Box<\[Value\]>? Do we want two layers of indirection when getting locals?
+    pub locals: Locals,
```

Comment:
This comments should be resolved

Reply:
I think we can leave it as is. Currently `Locals` is a one-field struct over a `Box<[Value]>`, so I would expect the compiler to not actually introduced a second level of indirection.

Reply:
I've noticed that we rely a lot on the compiler doing smart optimizations, yet we never check for it explicitly. Perhaps we should keep track of our assumptions, and eventually verify them (if we can?)

Reply:
That definitely is a good though. Though there are so many micro decisions that I would like refactor anyhow once we have the base set of features, that I'm not sure if it makes sense to spent the effort on tracking our assumptions.

One way of doing this could be writing unit tests that just assert the size of types, i.e. in this case the size of `Locals` to be that of an `usize`, or more precisely of a `Box<[ValTypes]>`.

Reply:
I agree with you that it is early. I suggest we start investigating this matter further after we have stabilized our architecture